### PR TITLE
HttpResponseHeaders validates trailing headers by names

### LINKED ISSUE #41262: HttpResponseMessage.TrailingHeaders blocks some headers
HttpResponseMessage.TrailingHeaders re-used the HttpResponseHeaders collection that blocks some headers it thinks belong in HttpContentHeaders instead. With trailers you don't have that flexibility and this may block customer scenarios.

A header like `Content-MD5` is restricted.
- The header is received and stored in the collection.
- `response.TrailingHeaders.ToString()` does show the value.
- `response.TrailingHeaders.GetValues("Content-MD5");` throws:
```
System.InvalidOperationException: Misused header name. Make sure request headers are used with HttpRequestMessage, response headers with HttpResponseMessage, and content headers with HttpContent objects.
at System.Net.Http.Headers.HttpHeaders.GetHeaderDescriptor(String name)
at System.Net.Http.Headers.HttpHeaders.GetValues(String name)
at ResponseTrailers.Program.Main(String[] args) in C:\temp\ResponseTrailers\Program.cs:line 30
```
- `response.TrailingHeaders.TryGetValues("Content-MD5", out var values)` returns false.
- Enumeration is the only way to get the value:
```
foreach (var header in response.TrailingHeaders)
{
    Console.WriteLine($"Trailer {header.Key}: {string.Join(", ", header.Value)}");
}
```

I reviewed the content headers that would be blocked and didn’t find any that people are likely to use, but devs are unpredictable. 

Proposed change: HttpResponseMessage owns and creates the trailers collection. It could be changed to create it using a new internal constructor that specified different behavior. No public API changes would be required. Note there are plenty of headers that are prohibited in trailers that could be filtered by these rules instead of the content headers.
https://github.com/dotnet/corefx/blob/9161098ade5075a8c126b2d7eb98abbeadff2961/src/System.Net.Http/src/System/Net/Http/HttpResponseMessage.cs#L123-L126
https://github.com/dotnet/corefx/blob/9161098ade5075a8c126b2d7eb98abbeadff2961/src/System.Net.Http/src/System/Net/Http/Headers/HttpResponseHeaders.cs#L144-L145

The secondary issue is that HttpResponseHeaders has many properties that would not apply to trailers. That is a long term API wart, but not blocking.

.NET Core 3.0 based repro. Run Kestrel directly, not IIS Express.
[ResponseTrailers.zip](https://github.com/dotnet/corefx/files/3643612/ResponseTrailers.zip)

### PR DESCRIPTION
**Pull request by alnikola**
HttpResponseMessage.TrailingHeaders property is set to a new special instance of HttpResponseHeaders configured to validate trailing headers based on their names instead of types. When a header is being added via an explicit Add call, it checks the header name against prohibited trailing header list and throws an exception if it is found. In contrast, when a header arrives on the wire and is being added with TryAddWithoutValidation call, no extra check is performed because HttpConnection validates trailing headers in this scenario.
Fixes #41262

### CODE REVIEW

File: src/System.Net.Http/src/System/Net/Http/HttpResponseMessage.cs [hunk 1]
```diff
@@ -122,7 +122,7 @@ public HttpResponseHeaders TrailingHeaders
             {
                 if (_trailingHeaders == null)
                 {
-                    _trailingHeaders = new HttpResponseHeaders();
+                    _trailingHeaders = new HttpResponseHeaders(true);
```
Comment (scalablecory) [inline src/System.Net.Http/src/System/Net/Http/HttpResponseMessage.cs]:
```suggestion
                    _trailingHeaders = new HttpResponseHeaders(containsTrailingHeaders: true);
```

File: src/System.Net.Http/src/System/Net/Http/Headers/HttpResponseHeaders.cs [hunk 2]
```diff
@@ -159,6 +164,21 @@ internal override void AddHeaders(HttpHeaders sourceHeaders)
...
         }
 
+        internal override bool IsAllowedHeaderName(string headerName)
+        {
+            return !_containsTrailingHeaders || headerName != null && !ProhibitedTrailerHeaders.Contains(headerName);
+        }
+
+        private static HashSet<string> GetProhibitedTrailerHeaders()
+        {
+            var result = new HashSet<string>();
```
Comment (scalablecory) [inline src/System.Net.Http/src/System/Net/Http/Headers/HttpResponseHeaders.cs]: header names are case-insensitive, should there be a comparer here?

File: src/System.Net.Http/src/System/Net/Http/Headers/HttpResponseHeaders.cs [hunk 3]
```diff
@@ -17,9 +18,11 @@ public sealed class HttpResponseHeaders : HttpHeaders
         private const int VarySlot = 3;
         private const int WwwAuthenticateSlot = 4;
         private const int NumCollectionsSlots = 5;
+        private static HashSet<string> ProhibitedTrailerHeaders = GetProhibitedTrailerHeaders();
```
Comment (scalablecory) [inline src/System.Net.Http/src/System/Net/Http/Headers/HttpResponseHeaders.cs]:
```suggestion
        private static HashSet<string> s_prohibitedTrailerHeaders = GetProhibitedTrailerHeaders();
```

File: src/System.Net.Http/src/System/Net/Http/Headers/KnownHeaders.cs [hunk 4]
```diff
@@ -97,6 +99,32 @@ internal static class KnownHeaders
         public static readonly KnownHeader XRequestID = new KnownHeader("X-Request-ID");
         public static readonly KnownHeader XUACompatible = new KnownHeader("X-UA-Compatible");
 
+        public static readonly IReadOnlyList<KnownHeader> TrailerDisallowedHeaders = Array.AsReadOnly(new[]    // rfc7230 4.1.2.
```
Comment (scalablecory) [inline src/System.Net.Http/src/System/Net/Http/Headers/KnownHeaders.cs]: Can this be made a `HashSet` so we don't need the extra copy here and in `HttpConnection`?

File: src/System.Net.Http/src/System/Net/Http/Headers/HttpResponseHeaders.cs [hunk 5]
```diff
@@ -141,9 +144,11 @@ public HttpHeaderValueCollection<WarningHeaderValue> Warning
 
         #endregion
 
-        internal HttpResponseHeaders()
-            : base(HttpHeaderType.General | HttpHeaderType.Response | HttpHeaderType.Custom, HttpHeaderType.Request)
+        internal HttpResponseHeaders(bool containsTrailingHeaders = false)
+            : base(containsTrailingHeaders ? HttpHeaderType.All & (HttpHeaderType.All ^ HttpHeaderType.Request) : HttpHeaderType.General | HttpHeaderType.Response | HttpHeaderType.Custom,
```
Comment (scalablecory) [inline src/System.Net.Http/src/System/Net/Http/Headers/HttpResponseHeaders.cs]:
```suggestion
            : base(containsTrailingHeaders ? HttpHeaderType.All ^ HttpHeaderType.Request : HttpHeaderType.General | HttpHeaderType.Response | HttpHeaderType.Custom,
```

File: src/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/HttpConnection.cs [hunk 6]
```diff
@@ -42,31 +42,7 @@ internal partial class HttpConnection : HttpConnectionBase, IDisposable
...
-            KnownHeaders.Authorization, KnownHeaders.SetCookie,
-
-            // Response control data.
-            // rfc7231#section-7.1: Control Data.
-            KnownHeaders.Age, KnownHeaders.Expires, KnownHeaders.Date, KnownHeaders.Location, KnownHeaders.RetryAfter, KnownHeaders.Vary, KnownHeaders.Warning,
-
-            // Content-Encoding, Content-Type, Content-Range, and Trailer itself.
-            KnownHeaders.ContentEncoding, KnownHeaders.ContentType, KnownHeaders.ContentRange, KnownHeaders.Trailer
-        };
+        private static readonly HashSet<KnownHeader> s_disallowedTrailers = new HashSet<KnownHeader>(KnownHeaders.TrailerDisallowedHeaders);
```
Comment (scalablecory) [inline src/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/HttpConnection.cs]: Trailing headers are so rare that this probalby doesn't matter, but if we added a `HttpHeaderType.NonTrailing`, we could get rid of this `HashSet` all together. It is only used in one place:

https://github.com/dotnet/corefx/blob/3d724ba3038b4d439038364f898010e9e4189e21/src/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/HttpConnection.cs#L902-L908

File: src/System.Net.Http/src/System/Net/Http/Headers/KnownHeaders.cs (hunk 4 above)
Comment (alnikola) [reply src/System.Net.Http/src/System/Net/Http/Headers/KnownHeaders.cs]: TrailerDisallowedHeader is a public static field whose contents must not be changed, so it's better to have a readonly collection here to avoid accidental modifications. Unfortunately, there is no IReadOnlySet interface, so I decided to use IReadOnlyList.

File: src/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/HttpConnection.cs (hunk 6 above)
Comment (alnikola) [reply src/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/HttpConnection.cs]: I need a header name-based comparison in HttpResponseHeader because known headers might get converted to custom one which are compared by names. Thus, I cannot completely get rid of HashSet even if we introduce the new header type.

File: src/System.Net.Http/src/System/Net/Http/Headers/KnownHeaders.cs (hunk 4 above)
Comment (stephentoub) [reply src/System.Net.Http/src/System/Net/Http/Headers/KnownHeaders.cs]: It's public but on an internal class.  And there are plenty of other such things on this type where our own internal consuming code is expected to do the "right thing", e.g. if you look at KnownHeader, you'll see several exposed arrays where consumers shouldn't change the contents.  We should prioritize perf here.

Comment (alnikola) [reply src/System.Net.Http/src/System/Net/Http/Headers/KnownHeaders.cs]: Ok, I got the point, but anyways having a HashSet<KnownHeader> is not enough for my new logic in HttpResponseHeaders because known headers might get converted to custom ones which are compared only by names. Thus, I cannot simply re-use HashSet<KnownHeader>, but have to create a HashSet<string> with a custom comparer.

File: src/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/HttpConnection.cs (hunk 6 above)
Comment (scalablecory) [reply src/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/HttpConnection.cs]: You can use `KnownHeaders.TryGetKnownHeader` to do such a lookup already, so we'd be able to get rid of both hash sets.

File: src/System.Net.Http/src/System/Net/Http/Headers/KnownHeaders.cs (hunk 4 above)
Comment (scalablecory) [reply src/System.Net.Http/src/System/Net/Http/Headers/KnownHeaders.cs]: This list is only being used to initialize these two `HashSet`:

https://github.com/dotnet/corefx/blob/db05f70e41c4187e1ad44c700f078f45c1e9e3cb/src/System.Net.Http/src/System/Net/Http/Headers/HttpResponseHeaders.cs#L21

https://github.com/dotnet/corefx/blob/db05f70e41c4187e1ad44c700f078f45c1e9e3cb/src/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/HttpConnection.cs#L45

The ask here is that we either:
a) get rid of both hashsets and use a flag as suggested in https://github.com/dotnet/corefx/pull/42229/files#r340850684
b) if we want to keep both hashsets, use the `HashSet<KnownHeader>` to initialize the `HashSet<string>`.

And either way, we should get rid of this list.

File: src/System.Net.Http/src/System/Net/Http/Headers/HttpResponseHeaders.cs [hunk 7]
```diff
@@ -141,9 +144,11 @@ public HttpHeaderValueCollection<WarningHeaderValue> Warning
 
         #endregion
 
-        internal HttpResponseHeaders()
-            : base(HttpHeaderType.General | HttpHeaderType.Response | HttpHeaderType.Custom, HttpHeaderType.Request)
+        internal HttpResponseHeaders(bool containsTrailingHeaders = false)
+            : base(containsTrailingHeaders ? HttpHeaderType.All ^ HttpHeaderType.Request : HttpHeaderType.General | HttpHeaderType.Response | HttpHeaderType.Custom,
```
Comment (scalablecory) [inline src/System.Net.Http/src/System/Net/Http/Headers/HttpResponseHeaders.cs]: Enabling `HttpHeaderType.Content` here will cause `TryAddWithoutValidation("Content-Length", "123")` and `GetValues("Content-Length")` to succeed where they will currently fail for the trailing headers collection.

Comment (scalablecory) [reply src/System.Net.Http/src/System/Net/Http/Headers/HttpResponseHeaders.cs]: Suggestion here might work to resolve this: https://github.com/dotnet/corefx/pull/42229/files#r340850684

Comment (alnikola) [reply src/System.Net.Http/src/System/Net/Http/Headers/HttpResponseHeaders.cs]: This is caused by the issue with an incomplete disallowed header list in HttpConnection I reported yesterday. Currently, I just re-use that list of prohibited headers, however it misses some headers mentioned in the RFC (e.g. Proxy-Authorization). Content-Length is just another example.
It seems introducing new HttpHeaderType.NonTrailing is a good option. However, I have one concern. Currently, HttpHeaderType enum have mutually exclusive fields with only one exception of All. If I add NonTrailing flag, it will be set in combination with others (e.g. Request and NonTrailing), and it won't be clear how All should behave in this case. It seems NonTrailing represents a different concept because it doesn't define a header type (i.e. splits the set of all existing header into non-intersecting subsets ), but instead specify a permission for the given header instance.
Don't you think it could cause problems in the future? How will we treat All value? As "any header type" & "non in trailer" or as "any header" & "in trailer"?

Comment (scalablecory) [reply src/System.Net.Http/src/System/Net/Http/Headers/HttpResponseHeaders.cs]: `All` is only used in a single place right now, so I'm not really worried about it.

However there are places in code that do an explicit e.g. `if(headerType == HttpHeaderType.Response)` that would need to be changed to a `HasFlag`. I don't see this as a problem.

Comment (alnikola) [reply src/System.Net.Http/src/System/Net/Http/Headers/HttpResponseHeaders.cs]: Even though ```All``` is use only in one place now,  the introduction of NonTrailing flag still breaks its abstraction. Let alone the possibility that it can be used wider later.

Comment (alnikola) [reply src/System.Net.Http/src/System/Net/Http/Headers/HttpResponseHeaders.cs]: @stephentoub What do you think about the above proposal to introduce NonTrailing flag and the concerns around it?

Comment (scalablecory) [reply src/System.Net.Http/src/System/Net/Http/Headers/HttpResponseHeaders.cs]: This way seems no-cost from a perf perspective, but I'm not tied to it if we can think of a better way.

I don't understand what you mean by abstraction -- do you mean you wouldn't want to see the enum used as flags? It already is, but perhaps changing the name to `HttpHeaderFlags` would make things clearer.

File: src/System.Net.Http/src/System/Net/Http/Headers/HttpHeaders.cs [hunk 8]
```diff
@@ -997,8 +997,15 @@ internal object GetParsedValues(HeaderDescriptor descriptor)
             return info.ParsedValue;
         }
 
+        internal virtual bool IsAllowedHeaderName(string headerName) => true;
+
         private void PrepareHeaderInfoForAdd(HeaderDescriptor descriptor, out HeaderStoreItemInfo info, out bool addToStore)
         {
+            if (!IsAllowedHeaderName(descriptor.Name))
+            {
+                throw new InvalidOperationException(SR.net_http_headers_not_allowed_header_name);
```
Comment (davidsh) [inline src/System.Net.Http/src/System/Net/Http/Headers/HttpHeaders.cs]: We should modify `SR.net_http_headers_not_allowed_header_name` so that the headerName shows up as part of the exception text.

Instead of:
"Misused header name. Make sure request headers are used with HttpRequestMessage, response headers with HttpResponseMessage, and content headers with HttpContent objects."

it would be better to see, i.e.
"Misused header name, 'MY-HEADER-NAME'. Make sure request headers are used with HttpRequestMessage, response headers with HttpResponseMessage, and content headers with HttpContent objects."

Doing this will greatly help with diagnostics.

You'll need to modify `SR.net_http_headers_not_allowed_header_name` to add a `{0}` in the resource string and fix up other places this string resource is used.

File: src/System.Net.Http/src/System/Net/Http/Headers/HttpResponseHeaders.cs (hunk 7 above)
Comment (davidsh) [reply src/System.Net.Http/src/System/Net/Http/Headers/HttpResponseHeaders.cs]: I think adding `HttpHeaderType.NonTrailing` is fine as it makes things very clear in the code. `All` isn't used much. And this is all internal code anyways. So, the priority should be performance and code maintenance.

File: src/System.Net.Http/src/System/Net/Http/Headers/KnownHeaders.cs (hunk 4 above)
Comment (davidsh) [reply src/System.Net.Http/src/System/Net/Http/Headers/KnownHeaders.cs]: >We should prioritize perf here

I agree with @stephentoub on this point. Since this is internal code, we should add more Assert's if we think it will help us call the code correctly.

File: src/System.Net.Http/src/System/Net/Http/Headers/HttpResponseHeaders.cs (hunk 7 above)
Comment (alnikola) [reply src/System.Net.Http/src/System/Net/Http/Headers/HttpResponseHeaders.cs]: @scalablecory An abstraction is a generalized description of some entities in the given set which defines a common set of rules and statements valid for all entities in the set. In the given case, ```All``` means "all headers which can appear in HTTP messages". If we introduce the ```NonTrailing``` flag to the same ```HttpHeaderType```, it breaks this statement because we have to explicitly set either 0 or 1 value for the respective bit in ```All``` definition which will lead to one of the following mutually exclusive definitions:

1. "all headers which can appear in HTTP messages and can be present in the trailer"
2. "all headers which can appear in HTTP messages and cannot be present in the trailer"

Comment (alnikola) [reply src/System.Net.Http/src/System/Net/Http/Headers/HttpResponseHeaders.cs]: That said, it seems an introduction of ```NonTrailing``` completely eliminates the necessity to have HashSet, so I change the implementation. However, it has another (more practical) disadvantage, now it's mandatory to use bit ```&``` for any header type comparison. Please, check the current PR version and let me know your opinion if the perf improvement really outweighs that inconvenience.

Comment (davidsh) [reply src/System.Net.Http/src/System/Net/Http/Headers/HttpResponseHeaders.cs]: @alnikola It looks like your latest commit has chosen (2) ""all headers which can appear in HTTP messages and cannot be present in the trailer""? Is that right?  If so, perhaps it makes sense to rename the `All` bit to something else like `AllNonTrailing`?

Comment (alnikola) [reply src/System.Net.Http/src/System/Net/Http/Headers/HttpResponseHeaders.cs]: That's correct, however it's only because I had no other choice but to assign either 0 or 1. I'm not sure what it's the best option. Overall, do you think that getting rid of HashSet really justifies that ambiguity and the requirement to always compare header types through the bitwise ```&```?

Comment (scalablecory) [reply src/System.Net.Http/src/System/Net/Http/Headers/HttpResponseHeaders.cs]: Since `All` is only used in one place, is it reasonable to get rid of it entirely?

Comment (scalablecory) [reply src/System.Net.Http/src/System/Net/Http/Headers/HttpResponseHeaders.cs]: What's the concern about using `&` to check for header type?

Comment (davidsh) [reply src/System.Net.Http/src/System/Net/Http/Headers/HttpResponseHeaders.cs]: >Please, check the current PR version and let me know your opinion if the perf improvement really outweighs that inconvenience.

Since this isn't a public API, the potential usability concerns for the "inconvenience" aren't very strong. So, the perf improvment does outweigh other concerns. I think we should either 1) Remove the `All` if not really needed, or 2) rename the `All` to something more descriptive. And there should be more Assert's added in the code to help with future code maintenance/refactoring.

Comment (alnikola) [reply src/System.Net.Http/src/System/Net/Http/Headers/HttpResponseHeaders.cs]: It seems there is a misunderstanding. My main concern is not about ```All```, but about the fact that I had to change all header type comparison expressions by introducing bit-wise ```&``` operator because now it's impossible to directly compare with via a simple == expression. This approach is really fragile because any new person unaware about this approach will do it wrong.
In example,

- Before ```descriptor.HeaderType == HttpHeaderType.Content```
- **Now** ```(descriptor.HeaderType & HttpHeaderType.Content) == HttpHeaderType.Content```

Please, check all the changes made in [my latest commit](https://github.com/dotnet/corefx/pull/42229/commits/2a762ea4f00c9cf264e810e7a9124d833af7db6a).

Comment (davidsh) [reply src/System.Net.Http/src/System/Net/Http/Headers/HttpResponseHeaders.cs]: >This approach is really fragile because any new person unaware about this approach will do it wrong.

I don't think it's fragile. I do agree that it is a change from how it worked before. But basically these are now working as proper bit flags which do require '&' operators to verify that the bit is set.

Perhaps you could write some helper methods (which marked correctly with attributes would be inline-able by the compiler).

To help a new person get it right, just add some Debug.Assert's in the right places. Those will help make sure that future coding changes adhere to the new pattern.

File: src/System.Net.Http/tests/UnitTests/Headers/HttpResponseHeadersTest.cs [hunk 9]
```diff
@@ -596,6 +634,27 @@ public void Trailer_ReadAndWriteProperty_CallsForwardedToHttpGeneralHeaders()
             Assert.Equal(2, headers.Trailer.Count);
         }
 
+        [Theory, MemberData(nameof(AllowedTrailingHeaders))]
+        public void Trailer_AddAndGetAllowedContentHeader_Success(string name, string value)
+        {
+            var trailingHeaders = new HttpResponseHeaders(true);
```
Comment (scalablecory) [inline src/System.Net.Http/tests/UnitTests/Headers/HttpResponseHeadersTest.cs]:
```suggestion
            var trailingHeaders = new HttpResponseHeaders(containsTrailingHeaders: true);
```

File: src/System.Net.Http/tests/UnitTests/Headers/HttpResponseHeadersTest.cs [hunk 10]
```diff
@@ -596,6 +634,27 @@ public void Trailer_ReadAndWriteProperty_CallsForwardedToHttpGeneralHeaders()
...
+            var actualHeaderValues = trailingHeaders.GetValues(name).ToArray();
+            Assert.Equal(1, actualHeaderValues.Length);
+
+            Assert.Equal(value, actualHeaderValues[0]);
+        }
+
+        [Theory, MemberData(nameof(ProhibitedTrailingHeaders))]
+        public void Trailer_AddProhibitedHeader_ThrowsException(string header)
+        {
+            var trailingHeaders = new HttpResponseHeaders(true);
```
Comment (scalablecory) [inline src/System.Net.Http/tests/UnitTests/Headers/HttpResponseHeadersTest.cs]:
```suggestion
            var trailingHeaders = new HttpResponseHeaders(containsTrailingHeaders: true);
```

File: src/System.Net.Http/src/System/Net/Http/Headers/HttpResponseHeaders.cs (hunk 7 above)
Comment (scalablecory) [reply src/System.Net.Http/src/System/Net/Http/Headers/HttpResponseHeaders.cs]: Thanks.

File: src/System.Net.Http/src/System/Net/Http/Headers/HttpResponseHeaders.cs [hunk 11]
```diff
@@ -141,9 +143,11 @@ public HttpHeaderValueCollection<WarningHeaderValue> Warning
...
```
Comment (scalablecory) [inline src/System.Net.Http/src/System/Net/Http/Headers/HttpResponseHeaders.cs]: `All` includes `NonTrailing`, but is used when `contrainsTrailingHeaders = true`.

File: src/System.Net.Http/src/System/Net/Http/Headers/KnownHeaders.cs (hunk 4 above)
Comment (scalablecory) [reply src/System.Net.Http/src/System/Net/Http/Headers/KnownHeaders.cs]: Thanks.

File: src/System.Net.Http/src/System/Net/Http/Headers/HttpHeaders.cs [hunk 12]
```diff
@@ -1091,7 +1098,7 @@ private HeaderDescriptor GetHeaderDescriptor(string name)
                 return descriptor.AsCustomHeader();
             }
 
-            throw new InvalidOperationException(SR.net_http_headers_not_allowed_header_name);
+            throw new InvalidOperationException(string.Format(SR.net_http_headers_not_allowed_header_name, name));
         }
 
         private bool TryGetHeaderDescriptor(string name, out HeaderDescriptor descriptor)
```
Comment (scalablecory) [inline src/System.Net.Http/src/System/Net/Http/Headers/HttpHeaders.cs]: Due to the `!= 0`, these two lines will unintentionally test true if `descriptor.HeaderType` and `_allowHeaderTypes` containing `NonTrailing`.

https://github.com/dotnet/corefx/blob/7eca2f0f6c51ff517657c943d602cca4b81fddf4/src/System.Net.Http/src/System/Net/Http/Headers/HttpHeaders.cs#L1092

https://github.com/dotnet/corefx/blob/7eca2f0f6c51ff517657c943d602cca4b81fddf4/src/System.Net.Http/src/System/Net/Http/Headers/HttpHeaders.cs#L1117

File: src/System.Net.Http/src/System/Net/Http/Headers/HttpResponseHeaders.cs (hunk 11 above)
Comment (alnikola) [reply src/System.Net.Http/src/System/Net/Http/Headers/HttpResponseHeaders.cs]: It doesn't matter in this case because all mask based header type checks must also be changed to accommodate for two variations (with NonTrailing flag and without) of the given header type value under the consideration. See my comment below.

File: src/System.Net.Http/src/System/Net/Http/Headers/HttpHeaders.cs (hunk 12 above)
Comment (alnikola) [reply src/System.Net.Http/src/System/Net/Http/Headers/HttpHeaders.cs]: Valid point and to fix it I had to introduce more bit-wise ```&``` operators. In example, the code above now looks like 
```if (((descriptor.HeaderType & ~HttpHeaderType.NonTrailing) & _allowedHeaderTypes) != 0)```
where ```_allowedHeaderTypes``` is initialized as follows
```_allowedHeaderTypes = allowedHeaderTypes & ~HttpHeaderType.NonTrailing```
It seems really awkward, so I'd like to again raise my question on whether we really want to go with NonTrailing flag approach. Is it really worth the performance gain we get here? Trailing headers are rare.

Comment (davidsh) [reply src/System.Net.Http/src/System/Net/Http/Headers/HttpHeaders.cs]: >Is it really worth the performance gain we get here? Trailing headers are rare.

@stephentoub already [commented](https://github.com/dotnet/corefx/pull/42229#discussion_r341194878) that we should prioritize performance.

I understand that some of the code might look akward due to the bit masking operations. But I previously suggested to create some helper methods that appropriately annotated would be inline-able by the compiler. That would make the code look cleaner.

Basically, the design change is here is moving from an enumeration describing the HeaderType to a bitmask describing the HeaderType. And bitmasks require various  '|' and '&' operations at times. Using helper methods to make the code look cleaner would help here.

Comment (alnikola) [reply src/System.Net.Http/src/System/Net/Http/Headers/HttpHeaders.cs]: In the given cases helper methods are not sufficient because the comparison works only when _allowedHeaderTypes and _treatAsCustomHeaderTypes are properly initialized. Of course, instead of the one time initialization we can repeat the operation ```& ~HttpHeaderType.NonTrailing``` for them as well in the helper method, but anyways it looks unnatural.
My point here that we are trying to optimize **a rare case** by trading it off for code cleanness and thus **increasing chance of bugs** in the future. I would agree if trailing headers were on the hot path, then it would be justified. However, it seems they are not.

Comment (davidsh) [reply src/System.Net.Http/src/System/Net/Http/Headers/HttpHeaders.cs]: @stephentoub can you comment on this? If we think the performance benefit is not that great then maybe the bitflag approach is not useful.

Comment (scalablecory) [reply src/System.Net.Http/src/System/Net/Http/Headers/HttpHeaders.cs]: If _`allowedHeaderTypes` is made to not contain `NonTrailing`, then the existing line `if ((descriptor.HeaderType & _allowedHeaderTypes) != 0)` does not need to be changed. Other comparisons, instead of having `if((headerType & foo) == foo)` can be made into `if((headerType & foo) != 0)` as well to avoid masking out `NonTrailing` from `headerType`.

Comment (alnikola) [reply src/System.Net.Http/src/System/Net/Http/Headers/HttpHeaders.cs]: @scalablecory Sorry, I didn't get your idea completely. Could you please explain how the proposed simplification can be applied to HttpHeaders.GetHeaderDescriptor and HttpHeaders.TryGetHeaderDescriptor methods?
In these 2 methods, checks must pass even when the given descriptor.HeaderType has NonTrailing flag. In example, if I just use ```if ((descriptor.HeaderType & _allowedHeaderTypes) != 0)``` the check will **fail for the Host** header which has **Request and NonTrailing flags**.
[In the current PR version](https://github.com/dotnet/corefx/blob/7d5a59fd67709575aa823b3ecb59a8b4b51f8412/src/System.Net.Http/src/System/Net/Http/Headers/HttpHeaders.cs#L1092) I use ```if (((descriptor.HeaderType & ~HttpHeaderType.NonTrailing) & _allowedHeaderTypes) != 0)``` where[ _allowedHeaderTypes is initialized](https://github.com/dotnet/corefx/blob/7d5a59fd67709575aa823b3ecb59a8b4b51f8412/src/System.Net.Http/src/System/Net/Http/Headers/HttpHeaders.cs#L57) as
```_allowedHeaderTypes = allowedHeaderTypes & ~HttpHeaderType.NonTrailing```. 

Could you please show how it will look in your version?

Comment (scalablecory) [reply src/System.Net.Http/src/System/Net/Http/Headers/HttpHeaders.cs]: > we are trying to optimize a *rare case*

Lets not get hung up on an optimization red herring. Bit flags being an optimization is just a value-add, not the goal. The issue here is that the PR was needlessly triplicating a collection.

Bit flags are a suggestion that can cleanly solve that, but I am not adamant that we use them. If we can avoid triplicating the collection in another way that we think is clearer and easier to use without impacting perf, I am fine with that.

> if trailing headers were on the hot path, then it would be justified.

Trailing headers are used heavily in gRPC, and that team has communicated a want for optimization from us. Header parsing is the single most expensive part of our code other than perhaps TLS, so we should make sure to look at it with a keen eye.

> increasing chance of bugs in the future.

Bit manipulation is fairly common in protocol work, so I don't think we should shy away from it. In my experience it is not particularly bug-prone, but we should by all means do what we can to make us more successful iterating this code in the future. Renaming `HttpHeaderType` to `HttpHeaderFlags` and implementing helper methods where appropriate as @davidsh suggests are two things worth exploring.

**Comment by maryamariyan** [comment]
Thank you for your contribution. As announced in dotnet/coreclr#27549 this repository will be moving to dotnet/runtime on November 13. If you would like to continue working on this PR after this date, the easiest way to move the change to dotnet/runtime is:
1.	In your corefx repository clone, create patch by running `git format-patch origin`
2.	In your runtime repository clone, apply the patch by running `git apply --directory src/corefx <path to the patch created in step 1>`

File: src/System.Net.Http/src/System/Net/Http/Headers/HttpHeaders.cs (hunk 12 above)
Comment (scalablecory) [reply src/System.Net.Http/src/System/Net/Http/Headers/HttpHeaders.cs]: > Could you please show how it will look in your version?

So the bug is that `if((descriptor.HeaderType & _allowedHeaderTypes) != 0)` decomposed into e.g.:

```c#
descriptor.HeaderType = Content | NonTrailing;
_allowedHeaderTypes = Request | NonTrailing;

int mask = descriptor.HeaderType & _allowedHeaderTypes; // mask is now NonTrailing.
if(mask != 0)
{
   // invalid true path: we wanted Request, not Content, but mask is != 0 due to testing for NonTrailing.
}
```

To fix this, we only need to remove `NonTrailing` from one side of that `&` expression. The change you made to `_allowedHeaderTypes` resolves the problem without making any other code changes:
```c#
_allowedHeaderTypes = allowedHeaderTypes & ~HttpHeaderType.NonTrailing;
```

Now that `if` decomposes into:

```c#
descriptor.HeaderType = Content | NonTrailing;
_allowedHeaderTypes = Request; // no longer testing for NonTrailing.

int mask = descriptor.HeaderType & _allowedHeaderTypes; // mask is now 0.
if(mask != 0)
{
   ...
}
```

File: src/System.Net.Http/src/System/Net/Http/Headers/KnownHeaders.cs [hunk 13]
```diff
@@ -2,6 +2,8 @@
 // The .NET Foundation licenses this file to you under the MIT license.
 // See the LICENSE file in the project root for more information.
 
+using System;
+using System.Collections.Generic;
```
Comment (davidsh) [inline src/System.Net.Http/src/System/Net/Http/Headers/KnownHeaders.cs]: Do you still need these 2 'using' statements now that you aren't doing `HashSet`?

Comment (scalablecory) [reply src/System.Net.Http/src/System/Net/Http/Headers/KnownHeaders.cs]: @stephentoub is there an analyzer we can enable for this?

Comment (stephentoub) [reply src/System.Net.Http/src/System/Net/Http/Headers/KnownHeaders.cs]: >is there an analyzer we can enable for this?

Yes, but at least from my experience it caused a fair amount of noise due to ifdefs and projects that built for multiple configurations, where one build would say a using wasn't needed but then another build would require it.  As such I've avoided turning it on.  A few months back I did a pass to run it, remove everything that wasn't needed, and manually put back the ones that then caused compilation failures.  We could do that again, or experiment again with turning on additional rules.

**Comment by alnikola** [comment]
I created a new issue for the remaining work #42474.
